clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -4248,6 +4277,184 @@ ExceptionSpecificationType Parser::ParseDynamicExceptionSpecification(
                                         Exceptions.empty());
   return Exceptions.empty() ? EST_DynamicNone : EST_Dynamic;
 }
+/// ParseFunctionContractSpecifierSeq - Parse a series of pre/post contracts on
+/// a
+///     function declaration.
+///
+///   function-contract-specifier-seq :
+///       function-contract-specifier function-contract-specifier-seq
+//
+///   function-contract-specifier:
+///       precondition-specifier
+///       postcondition-specifier
+//
+///   precondition-specifier:
+///       pre attribute-specifier-seq[opt] ( conditional-expression )
+///
+///   postcondition-specifier:
+///       post attribute-specifier-seq[opt] ( result-name-introducer[opt]
+///       conditional-expression )
+///
+///   result-name-introducer:
+///       attributed-identifier :
+void Parser::MaybeParseFunctionContractSpecifierSeq(
+    Declarator &DeclaratorInfo) {
+  ContractKeyword CKK;
+  while ((CKK = isContractSpecifier(Tok)) != ContractKeyword::None) {
+    StmtResult Contract = ParseFunctionContractSpecifier(DeclaratorInfo);
+    if (Contract.isUsable()) {
+      DeclaratorInfo.addContract(Contract.getAs<ContractStmt>());
+    }
+  }
+}
+
+StmtResult Parser::ParseFunctionContractSpecifier(Declarator &DeclaratorInfo) {
+  auto [CK, CKStr] = [&]() -> std::pair<ContractKind, const char *> {
+    switch (isContractSpecifier(Tok)) {
+    case ContractKeyword::Pre:
+      return std::make_pair(ContractKind::Pre, "pre");
+    case ContractKeyword::Post:
+      return std::make_pair(ContractKind::Post, "post");
+    default:
+      llvm_unreachable("unhandled case");
+    }
+  }();
+  SourceLocation KeywordLoc = Tok.getLocation();
+  ConsumeToken();
+
+  if (Tok.isNot(tok::l_paren)) {
+    Diag(Tok, diag::err_expected_lparen_after) << "contract_assert";
+    SkipUntil({tok::equal, tok::l_brace, tok::arrow, tok::kw_try, tok::comma,
+               tok::l_paren},
+              StopAtSemi | StopBeforeMatch);
+    return StmtError();
+  }
+
+  BalancedDelimiterTracker T(*this, tok::l_paren);
+  if (T.expectAndConsume(diag::err_expected_lparen_after, CKStr,
+                         tok::r_paren)) {
+    return StmtError();
+  }
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    if (CK != ContractKind::Post) {
+      // Only post contracts can have a result name
+      Diag(Tok.getLocation(), diag::err_expected) << tok::identifier;
+    }
+
+    IdentifierInfo *Id = Tok.getIdentifierInfo();
+    SourceLocation IdLoc = ConsumeToken();
+    (void)Id;
+    (void)IdLoc;
+    // FIXME(ericwf): Actually build the result name introducer
```

Comment:
@cor3ntin I'm struggling to get the return type out of the parser here. Thoughts?